felix/cmd/calico-bpf/commands/counters.go

```diff
@@ -0,0 +1,151 @@
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+// countersCmd represents the counters command
+var countersCmd = &cobra.Command{
+	Use:   "counters",
+	Short: "Show and reset counters",
+}
+
+func init() {
+	countersCmd.AddCommand(countersDumpCmd)
+	countersCmd.AddCommand(countersFlushCmd)
+	rootCmd.AddCommand(countersCmd)
+
+	countersDumpCmd.Flags().String("iface", "", "Interface name")
+	countersDumpCmd.Flags().Bool("all", false, "All interfaces")
+	countersFlushCmd.Flags().String("iface", "", "Interface name")
+	countersFlushCmd.Flags().Bool("all", false, "All interfaces")
+}
+
+var countersDumpCmd = &cobra.Command{
+	Use:   "dump",
+	Short: "dumps counters",
+	Run: func(cmd *cobra.Command, args []string) {
+		iface, all := parseFlags(cmd)
+		if all || iface == "" {
+			doForAllInterfaces(cmd, "dump", dumpInterface)
+		} else {
+			if err := dumpInterface(cmd, iface); err != nil {
+				log.WithError(err).Error("Failed to dump counter map.")
+			}
+		}
+	},
+}
+
+var countersFlushCmd = &cobra.Command{
+	Use:   "flush",
+	Short: "flush counters",
+	Run: func(cmd *cobra.Command, args []string) {
+		iface, all := parseFlags(cmd)
+		if all || iface == "" {
+			doForAllInterfaces(cmd, "flush", flushInterface)
+		} else {
+			if err := flushInterface(cmd, iface); err != nil {
+				log.WithError(err).Error("Failed to flush counter map.")
+			}
+		}
+	},
+}
+
+func parseFlags(cmd *cobra.Command) (string, bool) {
+	all, err := cmd.Flags().GetBool("all")
+	if err != nil {
+		log.WithError(err).Error("Failed to parse --all flag.")
+	}
+	iface, err := cmd.Flags().GetString("iface")
+	if err != nil {
+		log.WithError(err).Error("Failed to parse interface name.")
+	}
+	return iface, all
+}
+
+func doForAllInterfaces(cmd *cobra.Command, action string, fn func(cmd *cobra.Command, iface string) error) {
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		log.WithError(err).Error("failed to get list of interfaces.")
+		return
+	}
+	for _, i := range interfaces {
+		err = fn(cmd, i.Name)
+		if err != nil {
+			log.Errorf("Failed to %s interface %s", action, i.Name)
+			continue
+		}
+	}
+}
+
+func dumpInterface(cmd *cobra.Command, iface string) error {
+	if iface == "" {
+		return fmt.Errorf("empty interface name")
+	}
+
+	bpfCounters := counters.NewCounters(iface)
+	values, err := bpfCounters.Read()
+	if err != nil {
+		return fmt.Errorf("Failed to read bpf counters. iface=%s err=%v", iface, err)
+	}
+
+	table := tablewriter.NewWriter(os.Stdout)
+	table.SetHeader([]string{"", iface, "INGRESS", "EGRESS"})
+	genRow := func(description string, ingress, egress uint32) []string {
+		fields := strings.Split(description, " ")
+		category := fields[0]
+		caption := strings.Join(fields[1:], " ")
+		return []string{
+			category,
+			caption,
+			fmt.Sprintf("%v", ingress),
+			fmt.Sprintf("%v", egress),
+		}
+	}
+
+	var rows [][]string
+
+	noOfCounters := len(counters.Descriptions)
+	for c := 0; c < noOfCounters; c++ {
```

Comment:
I recommend sorting the values by description somehow.  Then, when we add a new counter it won't have to be stuck at the end.

Reply:
(If you're not sure how to do that, I can give you some pointers)

Reply:
Done. Is it OK now?